Refactor ArchiveInvalidator: dependency decoupling, responsibility limiting and code reorganization

diosmosis opened the pull request:
This PR includes the following changes:
- ArchiveInvalidator's dependency on SitesManager has been removed. The code that updates a sites created time has been moved to tracker, since it's reason for existence is to make sure if visits are tracked before a site's creation date, those visits will still be displayed in the UI. This code was moved to a new RequestProcessor located within the SitesManager plugin (so core is no longer dependent on SitesManager for this at least).
- The date parsing logic was moved from ArchiveInvalidator to callers. The ArchiveInvalidator service now takes well-formed input only. At least for the dates.
- Some code related to PrivacyManager dependency was grouped closer together in ArchiveInvalidator.php both physically and in terms of code execution.

Commit:
Remove depedency to SitesManager plugin in core ArchiveInvalidator service by moving logic to new request processor.

ArchiveInvalidator used to update site creation time. The reasoning was, if we tracked data in the past that was before
the site's creation time, the site creation time should be updated, so it will be displayed in the UI. The tracking
should trigger archive invalidation which should eventually trigger an update to the site creation time. This is
incorrectly mixing concerns. Now the new request processor will make this change at the moment old data is tracked.

New test included.

Commit:
Move date parsing logic from ArchiveInvalidator to CoreAdminHome/API.php so the service class only accepts well formed input.

Commit:
Replicate old logic in ArchiveInvalidator to the letter to get some tests to pass.

Commit:
In ArchiveInvalidator make date filtering an explicit step and group with logic to detect maximum date to purge from.

Commit:
Fix newly added test.

Commit:
Fixing failing system test (EcommerceOrderWithItemsTest).

tsteur commented on plugins/SitesManager/Tracker/SitesManagerRequestProcessor.php:
```diff
@@ -0,0 +1,72 @@
+    private function getSiteCreatedTime($idSite)
+    {
+        $attributes = Cache::getCacheWebsiteAttributes($idSite);
+        $tsCreated = isset($attributes['ts_created']) ? $attributes['ts_created'] : 0;
```
Returning `0` could result in the current day as `Date::factory(0) == now or today`. So we might update the sites day accidentally just because it was not in the cache. Do you know what I mean? If request time is earlier than now or today, the site will be updated. In reality the sites creation time could be like a couple of years back but because it's not in cache it would be suddenly set to the request timestamp?

tsteur replied:
Also not using `? : ` would sometimes improve readability (especially when looking at a diff side by side where there's not much space) but I guess that's just personal opinion so feel free to ignore :) I know many say exactly the other way around

diosmosis replied:
Ok, will handle this case specifically. And I don't have an issue w/ not using `? :` here.

tsteur commented on plugins/SitesManager/Tracker/SitesManagerRequestProcessor.php:
```diff
@@ -0,0 +1,72 @@
+    {
+        // if we successfully record some data and the date is older than the site's created time,
+        // update the created time so the data will be viewable in the UI
+        $visitorNotFoundInDb = $request->getMetadata('CoreHome', 'visitorNotFoundInDb');
```
When changing something in CoreHome and this metadata is no longer present etc, would we notice this in tests? I presume via system tests?

Also maybe it would be worth writing an integration test just for this class / method?

diosmosis replied:
I could try and add a new test to the test added in this PR. RequestProcessors are hard to test in isolation (if at all) since they depend on each other. An integration test that goes through tracking will catch regressions, testing the class by itself would be like testing w/ mocks in this case.

tsteur replied:
Yeah a unit test would be useless. An integration test could be written maybe kinda easy as dependencies are nicely injected from outside here and expected metadata and visit properties could be maybe set without having to mock anything? Not sure though, not so much into it right now :)

diosmosis replied:
Started writing a test for this then realized that this only happens on an existing visit, and a visit in the past will always be a new visit. After the new visit, it could occur, but at this point the site's created date will be earlier then the next action time. So I will remove the check entirely :)

tsteur commented on plugins/CoreAdminHome/API.php:
```diff
@@ -107,4 +114,37 @@ public function runCronArchiving()
+     * @param array $dates
+     * @return Date[]
+     */
+    private function getDatesToInvalidateFromString($dates)
```
Maybe this method could be moved somewhere else and made reusable? In the end you just give it some dates and you get back the invalid dates and the valid ones? Also some tests for this could be nice

diosmosis replied:
Right now just simplifying the invalidator to make it easier for me to change. Maybe I will create a DatesList class as we talked about re: https://github.com/piwik/piwik/issues/8852 if I end up w/ some time.

tsteur commented:
Looks good apart from the possible `Date::factory(0)`. Others are only suggestions.

Commit:
Cover case in SitesManagerRequestProcess where ts_created is not found in tracker cache more explicitly. Make sure the site created time is not modified in this case.

Commit:
Remove check for visitorNotFoundInDb request metadata since it will never be set.

If a tracker request is for an existing visit in the past, then the site creation time will always be earlier, since new action will be later than the original visit (so the original visit would have changed the creation time).
